server/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -257,51 +246,34 @@ public NodeEnvironment(Settings settings, Environment environment) throws IOExce
 
         try {
             sharedDataPath = environment.sharedDataFile();
-            IOException lastException = null;
-            int maxLocalStorageNodes = MAX_LOCAL_STORAGE_NODES_SETTING.get(settings);
+            IOException exception = null;
 
-            final AtomicReference<IOException> onCreateDirectoriesException = new AtomicReference<>();
-            for (int possibleLockId = 0; possibleLockId < maxLocalStorageNodes; possibleLockId++) {
-                try {
-                    nodeLock = new NodeLock(possibleLockId, logger, environment,
-                        dir -> {
-                            try {
-                                Files.createDirectories(dir);
-                            } catch (IOException e) {
-                                onCreateDirectoriesException.set(e);
-                                throw e;
-                            }
-                            return true;
-                        });
-                    break;
-                } catch (LockObtainFailedException e) {
-                    // ignore any LockObtainFailedException
-                } catch (IOException e) {
-                    if (onCreateDirectoriesException.get() != null) {
-                        throw onCreateDirectoriesException.get();
-                    }
-                    lastException = e;
-                }
+            try {
+                nodeLock = new NodeLock(logger, environment,
+                    dir -> {
+                        Files.createDirectories(dir);
```

Comment:
Can we create the directories first, and then this argument can become `p -> true`? Arguably we could replace it with `p -> true` everywhere and therefore drop it entirely.

Reply:
> Arguably we could replace it with p -> true everywhere and therefore drop it entirely

not quite. For the node tools, we just select the directories that exist (i.e. possibly a subset). Here, we want to enforce that they all exist.

Reply:
I think it'd be ok for the node tools to fail if `path.data` points to paths that don't exist, rather than skipping over them as they do today. You can always adjust `path.data` to get the tool to run.

Reply:
I'll mull over this, and open a follow-up PR